File: golang/two_pointer/is_subsequence.go
package main

func preprocess(t string) map[rune][]int {
	charIndices := make(map[rune][]int)

	for i, char := range t {
		charIndices[char] = append(charIndices[char], i)
	}

	return charIndices
}

func isSubsequenceOptimized(s string, t string) bool {
	charIndices := preprocess(t)

	currentIndex := -1

	for _, char := range s {
		indices, exists := charIndices[char]

		if !exists {
			return false
		}

		found := false

		for _, index := range indices {
			if index > currentIndex {
				currentIndex = index

				found = true

				break
			}
		}

		if !found {
			return false
		}
	}

	return true
}

File: python/two_pointer/is_subsequence.py
class Solution:
    def preprocess(self, t: str):
        char_indices = {}

        for index, char in enumerate(t):
            if char not in char_indices:
                char_indices[char] = []

            char_indices[char].append(index)

        return char_indices

    def isSubsequence(self, s: str, t: str) -> bool:
        char_indices = self.preprocess(t)

        current_index = -1

        for char in s:
            if char not in char_indices:
                return False

            found = False
            for index in char_indices[char]:
                if index > current_index:
                    current_index = index

                    found = True

                    break

            if not found:
                return False

        return True
